Feature: add bun.js runtime [WIP]

abdelhamiderrahmouni opened the pull request:
[WIP]
One of the most hot runtimes nowadays is bun, it is fast, reliable and for me i really need it my projects.

I have tried integrating it with vitodeploy and it works. But I am stuck in one problem I could not solve yet, which has to do with starting, pausing, and restarting the worker that launches the bun start script.
when i start it it stays in starting state, when i pause it same, and when i restart it it fails.

I would love to have some help from the interested folks here !
Thanks in advance ❤️

abdelhamiderrahmouni committed:
Feat: add bun.js runtime

mjsarfatti commented:
I'd love bun support! I haven't tested your PR yet, but have you tried removing the '&' from https://github.com/vitodeploy/vito/blob/e403fdae683c4eab95cdd3894100efc8e2a2025c/app/SiteTypes/Bun.php#L104 ?

abdelhamiderrahmouni commented:
Yes i have, the command at first was running without & and had the same problem. I have added it to try and solve the problem but didn't work.

abdelhamiderrahmouni commented:
Another thing i did't mention before is about deployments, they fail for some raison even after build has been successful
I think they might be linked.

<img width="1036" height="637" alt="image" src="https://github.com/user-attachments/assets/9412cb5f-1d8e-46cb-8e0b-59de086539b4" />

and i get the following logs when I run the "failed" deployements

```txt
[2025-12-17 11:44:06] local.INFO: shopt -s expand_aliases
alias php=/usr/bin/php
export SITE_PATH=/home/vito/bun.errahmounispace.com
export DOMAIN=bun.errahmounispace.com
export BRANCH=main
export REPOSITORY=abdelhamiderrahmouni/bun-nextjs
export COMMIT_ID=5bc55ce85027ef9fa511dc7387467c0b5c12605a
export PHP_VERSION=
export PHP_PATH=/usr/bin/php
if ! cd /home/vito/bun.errahmounispace.com; then
    echo 'VITO_SSH_ERROR' && exit 1
fi

cd $SITE_PATH

git pull origin $BRANCH

bun install

bun --bun run build

sudo supervisorctl restart all

echo "✅ Deployment completed successfully!"
  
[2025-12-17 11:44:30] local.ERROR: Error executing command {"msg":"SSH command failed with an error","log":{"App\\Models\\ServerLog":{"server_id":3,"name":"3-1765971868-restart-workers.log","type":"restart-workers","disk":"server-logs","site_id":1,"updated_at":"2025-12-17T11:44:28.000000Z","created_at":"2025-12-17T11:44:28.000000Z","id":131}}} 
```

abdelhamiderrahmouni committed:
removed `&` from the end of the start command in SiteTypes/Bun.php

abdelhamiderrahmouni commented:
Just found the problem.

The Installation of bun when running their script `curl -fsSL https://bun.sh/install | bash -s "bun-v${BUN_VERSION}"` goes to `/home/vito/.bun/bin/bun` which is not globally accessible especially for supervisor's workers so the fix is to either specify the full path to bun when running it or installing it Globally.

what should we do ?

abdelhamiderrahmouni commented on app/SiteTypes/Bun.php:
```diff
@@ -0,0 +1,172 @@
+            $this->site->id
+        );
+        $this->site->server->ssh()->exec(
+            __('bun --bun run --cwd :path build', [
+                'path' => $this->site->path,
+            ]),
+            'bun-build',
+            $this->site->id
+        );
+        $this->progress(65);
+        $command = __('bun --bun run --cwd :path start', [
```
the `--bun` flag is necessary as stated by the bun team because it forces bun to use itself instead of Nodejs for serving the application, this is especially true for Nextjs.

abdelhamiderrahmouni commented on resources/deployment-scripts/bun.sh:
```diff
@@ -0,0 +1,11 @@
+
+bun install
+
+bun --bun run build
```
no it is consistent, bun install is a particular command it does not need the `--bun` flag but the other commands like `run dev`, `run build` and `run start` need the `--bun` flag.

abdelhamiderrahmouni committed:
Fix: `install` in `bun install` should be before `--cwd :path` and not after it because it is different than `bun run dev`, `bun run build` ...etc

abdelhamiderrahmouni committed:
refactor: change Bun installation to install bun globally to /usr/local/bin instead of per user and clean up PATH entries.

abdelhamiderrahmouni commented:
I have just changed the installation to become global

added
```bash
# Move Bun binary to /usr/local/bin for global access
sudo mv ~/.bun/bin/bun /usr/local/bin/
sudo chmod a+x /usr/local/bin/bun

# The installation script adds  "~/.bun/bin" to $PATH in "~/.bashrc" automatically, we should remove it
sed -i '/\.bun\/bin/d' "$HOME/.bashrc" 2>/dev/null || true
sed -i '/\.bun\/bin/d' "$HOME/.zshrc" 2>/dev/null || true
```

and the bun worker works well now!

abdelhamiderrahmouni commented:
Installation still errors in `isolated user` case

```txt
bun install v1.3.4 (5eb2145b)

EACCES: Permission denied: could not create the "node_modules" directory (mkdir)
```

abdelhamiderrahmouni committed:
Fix: add site->user to ssh methode to enable the installation for isolate mode.

abdelhamiderrahmouni commented:
My bad I was using $this->site->server->ssh()->exec(...) to execute the install and build for bun dependencies, i had to use $this->site->server->ssh($this->site->user)->exec(...) to scope the execution to the isolated user. 🤦‍♂️

Now everything works well, just like with NodeJS.

saeedvaziry committed:
Merge branch '3.x' into 3.x

saeedvaziry committed:
fix code style

abdelhamiderrahmouni committed:
Fix: move bunx binary to /usr/local/bin and update permissions for global access

saeedvaziry commented:
Since this is a plugin now, shall we close the PR?

abdelhamiderrahmouni commented:
Yes of course, Thank you so much @saeedvaziry .
